### PR TITLE
Refactor charts controller

### PR DESCRIPTION
**Pull request by mikestone14**
This PR refactors the charts controller to extract logic to various other modules, such as services and models.

### CODE REVIEW

File: server/services/BarChartTableService/BarChartTableService.test.js [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+      expect(service.csvTableData()).toEqual({
+        tableColumns: [SITE, 'Good', 'Bad', TOTAL_LABEL],
+        tableRows: [
+          ['Site 1', '1', '2', '3'],
+          ['Site 2', '4 / 4 (100%)', '6 / 6 (100%)', '10'],
+        ],
+      })
+    })
+  })
```
Comment (Elkrival) [inline server/services/BarChartTableService/BarChartTableService.test.js:94]: What do you think about adding a test that implements the Totals row and it checks that it's sorted properly..?

File: server/routes/charts.js [hunk 2]
```diff
@@ -339,9 +220,9 @@ router
 
       return updatedChart.modifiedCount === 1
         ? res.status(200).json({ data: result })
-        : res.status(404).json({ message: 'Chart could not be shared' })
+        : res.status(404).json({ message: "Chart could not be shared" })
     } catch (error) {
-      return res.status(404).json({ message: 'Chart could not be shared' })
+      return res.status(404).json({ message: "Chart could not be shared" })
```
Comment (Elkrival) [inline server/routes/charts.js]: We might have to check our editor settings and standardize this :)

File: server/data_processors/BarChartDataProcessor/index.js [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+
+      if (existingTotalsData) {
+        this.dataMap.set(totalsDataKey, existingTotalsData + variableCount)
+      } else {
+        this.dataMap.set(totalsDataKey, variableCount)
+      }
+    } else {
+      if (!this.dataMap.get(dataKey)) {
+        this.dataMap.set(dataKey, 0)
+      }
```
Comment (Elkrival) [inline server/data_processors/BarChartDataProcessor/index.js]:
```suggestion
      if (!this.dataMap.get(dataKey))  this.dataMap.set(dataKey, 0)
      
```

File: server/data_processors/BarChartDataProcessor/index.js [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    const totalsValueTargets = {}
+
+    for (const [key, count] of this.dataMap) {
+      const [study, valueLabel, targetValue] = key.split('-')
+      const totalsForStudy = this.studyTotals[study]
+      const totals = totalsForStudy.targetTotal || totalsForStudy.count
+      const percent = studyCountsToPercentage(count, totals)
+      const existingEntriesForStudy = processedDataBySite.get(study)
+      const targetValueAsNumber = +targetValue
+      const targetValueIsNan = Number.isNaN(targetValueAsNumber)
```
Comment (Elkrival) [inline server/data_processors/BarChartDataProcessor/index.js]:
```suggestion
      const targetValueIsNaN = Number.isNaN(targetValueAsNumber)
```

File: server/configs/basePathConfig.js [hunk 5]
```diff
@@ -0,0 +1,3 @@
+const basePathConfig = ''
```
Comment (Elkrival) [inline server/configs/basePathConfig.js:1]: This file was in the gitignore before. unsure how it got in here :)

File: server/models/SubjectModel/index.js [hunk 6]
```diff
@@ -0,0 +1,136 @@
+import {
```
Comment (Elkrival) [inline server/models/SubjectModel/index.js:1]: Can you help me understand why these are called models..?
My thoughts are around Prisma and how the schemas are marked with the model key word...

File: server/models/SubjectModel/index.js [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+  SOCIODEMOGRAPHICS_FORM,
+  ALL_SUBJECTS_MONGO_PROJECTION,
+  STUDIES_TO_OMIT,
+} from '../../constants'
+import { collections } from '../../utils/mongoCollections'
+import FiltersService from '../../services/FiltersService'
+
+export const intersectSubjectsFromFilters = (filters) => {
+  const intersectedSubjects = new Map()
+  const [filterListA, ...rest] = Object.values(filters).map((filter) => {
```
Comment (Elkrival) [inline server/models/SubjectModel/index.js]:
```suggestion
  const [filteredSubjectsListA, ...rest] = Object.values(filters).map((filter) => {
```

File: server/models/SubjectModel/index.js [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+
+    return filterMap
+  })
+  filterListA.forEach((subjectData) => {
+    const { subject, collection, filter } = subjectData
+    intersectedSubjects.set(subject, [{ collection, filter }])
+
+    rest.forEach((filterList) => {
+      const existingSubjectData = intersectedSubjects.get(subject)
+      const filterSubjectData = filterList.get(subject)
```
Comment (Elkrival) [inline server/models/SubjectModel/index.js]:
```suggestion
      const filteredSubjectData = filterList.get(subject)
```

**Review by Elkrival — APPROVED** [review]
Some suggestions with naming and I think we need a test.

File: server/configs/basePathConfig.js (hunk 5 above)
Comment (mikestone14) [reply server/configs/basePathConfig.js:1]: we need it in order for the tests that render the pages to work so I removed this file from the gitignore.

File: server/models/SubjectModel/index.js (hunk 6 above)
Comment (mikestone14) [reply server/models/SubjectModel/index.js:1]: these models aren't tied to the database or anything, they are meant to hold business logic for a particular entity like subjects. This way we can have that logic consolidated in a model object vs spread around in utility functions or something like that. Does that make sense?

File: server/routes/charts.js (hunk 2 above)
Comment (mikestone14) [reply server/routes/charts.js]: oh yeah I think my new computer had different settings. I'll fix this!

File: server/services/BarChartTableService/BarChartTableService.test.js (hunk 1 above)
Comment (mikestone14) [reply server/services/BarChartTableService/BarChartTableService.test.js:94]: I think the existing tests have the totals row in the proper order. Removing the totals row or changing the order results in failing tests so I think we're good.

File: server/models/SubjectModel/index.js (hunk 6 above)
Comment (Elkrival) [reply server/models/SubjectModel/index.js:1]: Yes, that makes sense.  Thank you for explaining that 👍

File: server/controllers/Untitled-1.js [hunk 9]
```diff
@@ -0,0 +1,352 @@
...
+    OR: {
+      count: 1,
+      targetTotal: 10,
+    },
+    NC: {
+      count: 2,
+      targetTotal: 10,
+    },
+  },
+}
```
Comment (Elkrival) [inline server/controllers/Untitled-1.js]: This file is labeled as Untitle-1.js..? does it belong to fixtures..?

Comment (mikestone14) [reply server/controllers/Untitled-1.js]: lol whoops - no this is not supposed to be checked in. I'll remove it!

File: server/controllers/chartsController/chartsController.test.js [hunk 10]
```diff
@@ -0,0 +1,199 @@
...
+      })
+    })
+
+    describe('when unsuccessful', () => {
+      it('returns a 500 error with the error message', async () => {
+        const request = {}
+        const response = createResponse()
+
+        await chartsController.index(request, response)
+        expect(response.status).toHaveBeenCalledWith(500)
```
Comment (Elkrival) [inline server/controllers/chartsController/chartsController.test.js]:
```suggestion
        await chartsController.index(request, response)
        
        expect(response.status).toHaveBeenCalledWith(500)
```

File: server/controllers/chartsController/chartsController.test.js [hunk 11]
```diff
@@ -0,0 +1,199 @@
...
+      })
+    })
+
+    describe('when unsuccessful', () => {
+      it('returns a 500 error with the error message', async () => {
+        const request = {}
+        const response = createResponse()
+
+        await chartsController.new(request, response)
+        expect(response.status).toHaveBeenCalledWith(500)
```
Comment (Elkrival) [inline server/controllers/chartsController/chartsController.test.js]:
```suggestion
        await chartsController.new(request, response)
        
        expect(response.status).toHaveBeenCalledWith(500)
```

File: server/controllers/chartsController/chartsController.test.js [hunk 12]
```diff
@@ -0,0 +1,199 @@
...
+      })
+    })
+
+    describe('when unsuccessful', () => {
+      it('returns a 500 error with the error message', async () => {
+        const request = {}
+        const response = createResponse()
+
+        await chartsController.edit(request, response)
+        expect(response.status).toHaveBeenCalledWith(500)
```
Comment (Elkrival) [inline server/controllers/chartsController/chartsController.test.js]:
```suggestion
        await chartsController.edit(request, response)
        
        expect(response.status).toHaveBeenCalledWith(500)
```

File: server/controllers/chartsController/index.js [hunk 13]
```diff
@@ -1,214 +1,137 @@
...
-    })
+import { userFromRequest } from '../../utils/userFromRequestUtil'
+import chartsListPage from '../../templates/Chart.template'
+import newChartPage from '../../templates/NewChart.template'
+import viewChartPage from '../../templates/ViewChart.template'
+import editChartPage from '../../templates/EditChart.template'
+import { DEFAULT_CHART_FILTERS } from '../../constants'
+import SubjectModel from '../../models/SubjectModel'
+import BarChartService from '../../services/BarChartService'
+import BarChartTableService from '../../services/BarChartTableService'
```
Comment (Elkrival) [inline server/controllers/chartsController/index.js:13]:
```suggestion
import { userFromRequest } from '../../utils/userFromRequestUtil'
import chartsListPage from '../../templates/Chart.template'
import newChartPage from '../../templates/NewChart.template'
import viewChartPage from '../../templates/ViewChart.template'
import editChartPage from '../../templates/EditChart.template'
import { DEFAULT_CHART_FILTERS } from '../../constants'
import CsvService from '../../services/CSVService'
import SubjectModel from '../../models/SubjectModel'
import BarChartService from '../../services/BarChartService'
import BarChartTableService from '../../services/BarChartTableService'
```

File: server/controllers/chartsController/index.js [hunk 14]
```diff
@@ -1,214 +1,137 @@
 import { ObjectId } from 'mongodb'
-import deepEqual from 'deep-equal'
+import qs from 'qs'
+
+import CsvService from '../../services/CSVService'
```
Comment (Elkrival) [inline server/controllers/chartsController/index.js]: I like how in the test files the services were grouped together. This import could be in that group
```suggestion
```

**Review by Elkrival — COMMENTED** [review]
left some formatting code comments, looks good so far 👍

**Pull request merged**